src/common/selftest/selftest_esp_update.cpp

```diff
@@ -0,0 +1,344 @@
+                    updateProgress();
+                }
+            } else {
+                fclose(opened_file);
+                ++current_file;
+                progress_state = current_file != firmware_set.end()
+                    ? esp_upload_action::Start_flash
+                    : esp_upload_action::Finish_and_reset;
+            }
+            break;
+        }
+        case esp_upload_action::Finish_and_reset:
+            log_info(Network, "ESP finished flashing");
+            phase = PhasesSelftest::ESP_passed;
+            esp_loader_flash_finish(true);
+            espif_flash_deinitialize();
+            notify_reconfigure();
+            progress_state = esp_upload_action::Finish_wait_user;
+            current_file = firmware_set.begin();
+            readCount = 0;
+            break;
+        case esp_upload_action::Finish_wait_user:
+            if (continue_pressed) {
+                progress_state = esp_upload_action::Done;
+            }
+            break;
+        case esp_upload_action::ESP_error:
+        case esp_upload_action::USB_error: {
+            phase = PhasesSelftest::ESP_failed;
+            esp_loader_flash_finish(false);
+            progress_state = esp_upload_action::Error_wait_user;
+            current_file = firmware_set.begin();
+            readCount = 0;
+            fclose(opened_file);
+            break;
+        }
+        case esp_upload_action::Error_wait_user:
+            if (continue_pressed) {
+                progress_state = esp_upload_action::Done;
+            }
+            break;
+        case esp_upload_action::Aborted:
+            progress_state = esp_upload_action::Done;
+            break;
+        case esp_upload_action::Done:
+            break;
+        }
+
+        //actualize status
+        status_encoder_union u;
+        u.status.phase = phase;
+        u.status.progress = progress;
+        u.status.current_file = current_file_no;
+        u.status.count_of_files = files_to_upload;
+        status = u.u32;
+    }
+}
+
+static void EspTask(void *pvParameters) {
+    ESPUpdate update((int)pvParameters);
+
+    update.Loop();
+
+    vTaskDelete(NULL); // kill itself
+}
+
+void update_esp(bool force) {
+
+    uint tasks_running_at_start = uxTaskGetNumberOfTasks();
+
+    TaskHandle_t xHandle = nullptr;
+
+    xTaskCreate(
+        EspTask,              // Function that implements the task.
+        "ESP UPDATE",         // Text name for the task.
+        512,                  // Stack size in words, not bytes.
+        (void *)((int)force), // Parameter passed into the task.
+        osPriorityNormal,     // Priority at which the task is created.
+        &xHandle);            // Used to pass out the created task's handle.
+
+    if (xHandle) {
+        FSM_Holder fsm_holder(ClientFSM::Selftest, 0);
+        status_t status;
+        status.Empty();
+
+        // wait until task kills itself
+        while (uxTaskGetNumberOfTasks() > tasks_running_at_start) {
```

Comment:
This is a very speculative way of checking the termination of a task... please do not rely on counting tasks like this. For now it may be good enough, but from a long term perspective this can result in a hard-to-spot error.

Reply:
I know.
I don't like creating that task at all. I think it should be moved to some existing wi-fi related task. This was just way faster.